AK/UFixedBigInt.h

```diff
@@ -17,474 +18,378 @@
+// are doing.
 template<typename T>
-requires(sizeof(T) >= sizeof(u64) && IsUnsigned<T>) class UFixedBigInt;
+constexpr inline size_t assumed_bit_size = 0;
+template<>
+constexpr inline size_t assumed_bit_size<IntegerWrapper<false>> = sizeof(int) * 8;
+template<size_t bit_size>
+constexpr inline size_t assumed_bit_size<UFixedBigInt<bit_size>> = bit_size;
+template<OneOf<bool, u8, u16, u32, u64, DoubleWord> T>
+constexpr inline size_t assumed_bit_size<T> = sizeof(T) * 8;
 
-// FIXME: This breaks formatting
-// template<typename T>
-// constexpr inline bool Detail::IsIntegral<UFixedBigInt<T>> = true;
+template<typename T>
+concept ConvertibleToUFixedInt = (assumed_bit_size<T> != 0);
 
 template<typename T>
-constexpr inline bool Detail::IsUnsigned<UFixedBigInt<T>> = true;
+concept UFixedInt = (ConvertibleToUFixedInt<T> && !IsSame<T, IntegerWrapper<false>>);
+
 template<typename T>
-constexpr inline bool Detail::IsSigned<UFixedBigInt<T>> = false;
+concept BuiltInUFixedInt = OneOf<T, bool, u8, u16, u32, u64, DoubleWord>;
 
 template<typename T>
-struct NumericLimits<UFixedBigInt<T>> {
-    static constexpr UFixedBigInt<T> min() { return 0; }
-    static constexpr UFixedBigInt<T> max() { return { NumericLimits<T>::max(), NumericLimits<T>::max() }; }
+concept NotBuiltInUFixedInt = UFixedInt<T> && !BuiltInUFixedInt<T>;
+
+// FIXME: This breaks formatting
+// template<typename T>
+// constexpr inline bool Detail::IsIntegral<UFixedBigInt<T>> = true;
+
+template<size_t bit_size>
+constexpr inline bool Detail::IsUnsigned<UFixedBigInt<bit_size>> = true;
+template<size_t bit_size>
+constexpr inline bool Detail::IsSigned<UFixedBigInt<bit_size>> = false;
+
+template<size_t bit_size>
+struct NumericLimits<UFixedBigInt<bit_size>> {
+    using T = UFixedBigInt<bit_size>;
+
+    static constexpr T min() { return T {}; }
+    static constexpr T max() { return --T {}; }
     static constexpr bool is_signed() { return false; }
 };
 
-template<Unsigned T>
-struct UFixedBigIntMultiplicationResult {
-    T low;
-    T high;
-};
+// ===== UFixedBigInt itself =====
+template<size_t bit_size>
+constexpr auto get_storage_of(UFixedBigInt<bit_size> const& value) { return value.m_data; }
+
+template<size_t bit_size, typename Storage>
+class UFixedBigInt {
+private:
+    template<size_t other_bit_size, typename OtherStorage>
+    friend class UFixedBigInt;
+
+    constexpr UFixedBigInt(ForOverwriteTag) { }
+
+    using RawLowUFixedBigInt = UFixedBigInt<bit_size / 2>;
+    using RawHighUFixedBigInt = UFixedBigInt<(bit_size + 1) / 2>;
+    using LowUFixedBigInt = Conditional<bit_size / 2 == 64, u64, RawLowUFixedBigInt>;
+    using HighUFixedBigInt = Conditional<(bit_size + 1) / 2 == 64, u64, RawHighUFixedBigInt>;
 
-template<typename T>
-requires(sizeof(T) >= sizeof(u64) && IsUnsigned<T>) class UFixedBigInt {
 public:
-    using R = UFixedBigInt<T>;
+    // TODO: make private
+    Storage m_data;
```

Comment:
you can just mark the `get_storage_of` function as a friend to solve this

Reply:
This is mostly a reminder for myself to hide it after getting formatting done)